### ISSUE TITLE
Support type hints `Required`, `NotRequired`

### ISSUE THREAD

**Issue by TiansuYu** [pydantic/pydantic#7712]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

If you are familiar with `TypedDict`, you will know about the type hints `typing.Required` and `typing.NotRequired` introduced by [PEP-665](https://peps.python.org/pep-0655/), mainly targeting for usage on `TypedDict`.

I know that since pydantic V2, the [nullable / optional / required field declaration](https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields)  has been sorted out, and as long as we adhere to the way declared in the doc page, we are good to go. But not every dev on my codebase will have time / mindset to read through it, and make sure their declaration is as intended. However, a type hint like `Required / NotRequired` will be make that much more explicit.

A change introduced by this will be minimal: For as any field declared as `NotRequired`, the field validator simply has to define a `None` as default, if no default is given. But it will makes the field specified in a code base dealing with dozens of model schemas much easier to read and being understood. 

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [X] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [ ] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [X] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
I thought about this, but akaik `Required` and `NotRequired`'s meanings are not "mainly targeting" TypedDict, they're only defined when used with TypedDict.

Also, `NotRequired` suggests a field may be missing completely - `AttributeError` if you try to access it, none none-able.

For those reasons I don't think it makes sense to extend the meaning of these types.

**Comment by Viicos** [comment]
A new `MISSING` sentinel was introduced in https://github.com/pydantic/pydantic/pull/11883 as an experimental feature, which I believe also fulfills this use case. It is available to be tested in the [2.12.0a1](https://github.com/pydantic/pydantic/releases/tag/v2.12.0a1) release.

It relies on the draft [PEP 661](https://peps.python.org/pep-0661/), introducing sentinels into the standard library (that is why this feature is marked as experimental).

`MISSING` was used as a name instead of `UNSET`, to avoid confusion with the current concept of "unset" fields (such fields are tracked _implicitly_ by Pydantic, and are only excluded when setting `exclude_unset=True`).

During serialization, any field with `MISSING` as a value is excluded from the output, and the sentinel doesn't appear in the JSON Schema, making it an implementation detail of your model, not exposed to external consumers.

```python
from pydantic import BaseModel
from pydantic.experimental.missing_sentinel import MISSING

class Configuration(BaseModel):
    timeout: int | None | MISSING = MISSING

conf = Configuration()

conf.model_dump()
# {}
Configuration.model_json_schema()['properties']['timeout']
#> {'anyOf': [{'type': 'integer'}, {'type': 'null'}], 'title': 'Timeout'}}
```

Sentinels are [special-cased by static type checkers](https://peps.python.org/pep-0661/#typing), meaning there is no need for any kind of `MissingType` in annotations. This feature is only supported by Pyright as of today.

See [the documentation](https://docs.pydantic.dev/dev/concepts/experimental/#missing-sentinel) for more details.

Stabilization of this feature can be tracked in [this issue](https://github.com/pydantic/pydantic/issues/12090).
